cli/receivers.go

```diff
@@ -0,0 +1,308 @@
+			return v.Receivers[i].Name < v.Receivers[j].Name
+		})
+
+		return v
+	}
+
+	// invalid keeps track of all invalid receiver configurations
+	var invalid []result
+	// jobs keeps track of all receivers that need to be sent test notifications
+	var jobs []job
+
+	for _, receiver := range c.Receivers {
+		integrations := buildReceiverIntegrations(receiver, tmpl, logger)
+		for _, integration := range integrations {
+			if integration.Error != nil {
+				invalid = append(invalid, result{
+					Receiver:    receiver,
+					Integration: &integration.Integration,
+					Error:       integration.Error,
+				})
+			} else {
+				jobs = append(jobs, job{
+					Receiver:    receiver,
+					Integration: &integration.Integration,
+				})
+			}
+		}
+	}
+
+	if len(invalid)+len(jobs) == 0 {
+		return nil, ErrNoReceivers
+	}
+
+	if len(jobs) == 0 {
+		return newTestReceiversResult(testAlert, invalid, now), nil
+	}
+
+	numWorkers := maxTestReceiversWorkers
+	if numWorkers > len(jobs) {
+		numWorkers = len(jobs)
+	}
+
+	resultCh := make(chan result, len(jobs))
+	jobCh := make(chan job, len(jobs))
+	for _, job := range jobs {
+		jobCh <- job
+	}
+	close(jobCh)
+
+	g, ctx := errgroup.WithContext(ctx)
+	for i := 0; i < numWorkers; i++ {
+		g.Go(func() error {
+			for job := range jobCh {
+				v := result{
+					Receiver:    job.Receiver,
+					Integration: job.Integration,
+				}
+				if _, err := job.Integration.Notify(notify.WithReceiverName(ctx, job.Receiver.Name), &testAlert); err != nil {
+					v.Error = err
+				}
+				resultCh <- v
+			}
+			return nil
+		})
+	}
+	g.Wait() // nolint
+	close(resultCh)
+
+	results := make([]result, 0, len(jobs))
+	for next := range resultCh {
+		results = append(results, next)
+	}
+
+	return newTestReceiversResult(testAlert, append(invalid, results...), now), nil
+}
+
+func newTestAlert(c TestReceiversParams, startsAt, updatedAt time.Time) types.Alert {
+	var (
+		defaultAnnotations = model.LabelSet{
+			"summary":          "Notification test",
+			"__value_string__": "[ metric='foo' labels={instance=bar} value=10 ]",
+		}
+		defaultLabels = model.LabelSet{
+			"alertname": "TestAlert",
+			"instance":  "Alertmanager",
+		}
+	)
+
+	alert := types.Alert{
+		Alert: model.Alert{
+			Labels:      defaultLabels,
+			Annotations: defaultAnnotations,
+			StartsAt:    startsAt,
+		},
+		UpdatedAt: updatedAt,
+	}
+
+	if c.Alert != nil {
+		if c.Alert.Annotations != nil {
+			for k, v := range c.Alert.Annotations {
+				alert.Annotations[k] = v
+			}
+		}
+		if c.Alert.Labels != nil {
+			for k, v := range c.Alert.Labels {
+				alert.Labels[k] = v
+			}
+		}
+	}
+
+	return alert
+}
+
+type ReceiverIntegration struct {
+	Integration notify.Integration
+	Error       error
+}
+
+// buildReceiverIntegrations builds a list of integration notifiers off of a
+// receiver config.
+func buildReceiverIntegrations(nc *config.Receiver, tmpl *template.Template, logger log.Logger) []ReceiverIntegration {
```

Comment:
This is copied and modified from https://github.com/prometheus/alertmanager/blob/ba8da18fb2b769ace00d270d677980c4d57310e7/cmd/alertmanager/main.go#L133

I acknowledge it is not ideal to have to maintain this twice, but I could not figure out the proper import without a cycle and the modifications enables the command to still test the valid integrations.